src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.CodeAnalysis.Extensibility;
+using Microsoft.VisualStudio.CodeAnalysis.RuleSets;
+using Microsoft.VisualStudio.Shell.Interop;
+using SonarLint.VisualStudio.Integration.NewConnectedMode;
+using SonarLint.VisualStudio.Integration.Resources;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public interface IProjectsRuleSetProvider
+    {
+        bool HasRuleSetWithSonarAnalyzerRules(string projectFilePath);
+    }
+
+    [Export(typeof(IProjectsRuleSetProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class ProjectsRuleSetProvider : IProjectsRuleSetProvider, IDisposable
+    {
+        private readonly IProjectSystemHelper projectSystemHelper;
+        private readonly IActiveSolutionTracker activeSolutionTracker;
+        private readonly IVsFileChangeEx fileChangeService;
+        private readonly IConfigurationProvider configurationProvider;
+        private readonly ILogger logger;
+
+        private readonly Dictionary<string, ProjectData> projectPathToCachedData = new Dictionary<string, ProjectData>();
+        private readonly Dictionary<string, FileChangeTracker> filesToTrack = new Dictionary<string, FileChangeTracker>();
+
+        private bool isDisposed;
+
+        [ImportingConstructor]
+        public ProjectsRuleSetProvider(IHost host)
+            : this(host.GetService<IProjectSystemHelper>(), host.GetService<SVsFileChangeEx, IVsFileChangeEx>(),
+                  host.GetMefService<IActiveSolutionTracker>(), InMemoryConfigurationProvider.Instance, host.Logger)
```

Comment:
At present the _InMemoryConfigurationProvider_ will only accurately track if we are in connected more or not, but I guess that's ok for this class.

Reply:
Yes we only care about being in the new connected mode.